src/algorithms/tracking/TrackPropagation.cc

```diff
@@ -67,66 +67,82 @@ namespace eicrecon {
-            track_points.push_back(result);
+            track_points.push_back(std::move(result));
         }
 
         return track_points;
     }
 
 
 
-    edm4eic::TrackSegment* TrackPropagation::propagateToSurfaceList(const eicrecon::TrackingResultTrajectory *traj,
-                                                                    std::vector<std::shared_ptr<Acts::Surface>> targetSurfaces) {
-      // start a mutable TrackSegment
-      edm4eic::MutableTrackSegment track_segment;
-      decltype(edm4eic::TrackSegmentData::length)      length       = 0;
-      decltype(edm4eic::TrackSegmentData::lengthError) length_error = 0;
-
-      // loop over projection-target surfaces
-      for(const auto& targetSurf : targetSurfaces) {
-
-        // project the trajectory `traj` to this surface
-        edm4eic::TrackPoint *point;
-        try {
-          point = propagate(traj, targetSurf);
-        } catch(std::exception &e) {
-          m_log->warn("<> Exception in TrackPropagation::propagateToSurfaceList: {}; skip this TrackPoint and surface", e.what());
-        }
-        if(!point) {
-          m_log->trace("<> Failed to propagate trajectory to this plane");
-          continue;
-        }
-
-        // logging
-        m_log->trace("<> trajectory: x=( {:>10.2f} {:>10.2f} {:>10.2f} )",
-            point->position.x, point->position.y, point->position.z);
-        m_log->trace("               p=( {:>10.2f} {:>10.2f} {:>10.2f} )",
-            point->momentum.x, point->momentum.y, point->momentum.z);
-
-        // update the `TrackSegment` length
-        // FIXME: `length` and `length_error` are currently not used by any callers, and may not be correctly calculated here
-        if(track_segment.points_size()>0) {
-          auto pos0 = point->position;
-          auto pos1 = std::prev(track_segment.points_end())->position;
-          auto dist = edm4eic::magnitude(pos0-pos1);
-          length += dist;
-          m_log->trace("               dist to previous point: {}", dist);
-        }
-
-        // add the `TrackPoint` to the `TrackSegment`
-        track_segment.addToPoints(*point);
-
-      } // end `targetSurfaces` loop
-
-      // output
-      track_segment.setLength(length);
-      track_segment.setLengthError(length_error);
-      return new edm4eic::TrackSegment(track_segment);
+    std::unique_ptr<edm4eic::TrackSegmentCollection> TrackPropagation::propagateToSurfaceList(
+        std::vector<const eicrecon::TrackingResultTrajectory*> trajectories,
+        std::vector<std::shared_ptr<Acts::Surface>> targetSurfaces
+        )
+    {
+      // logging
+      m_log->trace("Propagate trajectories: --------------------");
+      m_log->trace("number of trajectories: {}",trajectories.size());
+
+      // start output collection
+      auto track_segments = std::make_unique<edm4eic::TrackSegmentCollection>();
+
+      // loop over input trajectories
+      for(const auto& traj : trajectories) {
```

Comment:
I don't understand why do you prefer to move this loop inside the function. That restricts flexibility without any (performance?) gains.

Reply:
This is for modularity: less code in the factory, more code in the algorithm. Ideally the factories should just be boilerplate at this point, but in many cases this is still not 100% possible. This refactor does not restrict flexibility since input `trajectories` can have `size==1`.